lib/validate.js

```diff
@@ -0,0 +1,29 @@
+module.exports = validate
+
+var Promise = require('lie')
+
+function validate (state, doc) {
+  var self = this
+  return new Promise(function (resolve, reject) {
+    var result
+
+    try {
+      result = state.validate.call(self, doc)
+    } catch (error) {
+      reject(error)
+    }
+
+    if (result && result.then) {
+      return result.then(resolve, reject)
+    }
+
+    if (!result) {
+      var error = new Error('Document is not valid')
+
+      error.name = 'ValidationError'
+      reject(error)
+    }
+
+    resolve(doc)
```

Comment:
It would have worked if the validation was supposed to be triggered manually, but we are coupling it with CRUD methods isn't it?